subsys/console/getchar.c

```diff
@@ -6,53 +6,122 @@
 
 #include <zephyr.h>
 #include <device.h>
+#include <sys/ring_buffer.h>
 #include <console/console.h>
 #include <console/tty.h>
 #include <drivers/uart.h>
 
-static struct tty_serial console_serial;
+static struct device *uart_dev;
 
-static uint8_t console_rxbuf[CONFIG_CONSOLE_GETCHAR_BUFSIZE];
+#if defined(CONFIG_CONSOLE_RX_INTERRUPT_DRIVEN)
+
+static struct tty_serial console_serial;
 static uint8_t console_txbuf[CONFIG_CONSOLE_PUTCHAR_BUFSIZE];
+static uint8_t console_rxbuf[CONFIG_CONSOLE_GETCHAR_BUFSIZE];
+
+#else
+
+#define POLL_PERIOD K_MSEC(CONFIG_CONSOLE_POLLING_RX_POLL_PERIOD)
+
+void poll_timer_handler(struct k_timer *dummy);
+
+RING_BUF_DECLARE(console_rx_ringbuf, CONFIG_CONSOLE_GETCHAR_BUFSIZE);
+
+K_SEM_DEFINE(rx_sem, 0, CONFIG_CONSOLE_GETCHAR_BUFSIZE);
+
+K_TIMER_DEFINE(poll_timer, poll_timer_handler, NULL);
+
+
+void poll_timer_handler(struct k_timer *dummy)
+{
+	int ret = 0;
+	uint8_t c;
+
+	while (true) {
+		if (ring_buf_space_get(&console_rx_ringbuf) == 0) {
+			return;
+		}
+
+		ret = uart_poll_in(uart_dev, &c);
+		if (ret != 0) {
+			return;
+		}
+
+		ring_buf_put(&console_rx_ringbuf, &c, 1);
+
+		k_sem_give(&rx_sem);
+	}
+}
+
+#endif
+
 
 ssize_t console_write(void *dummy, const void *buf, size_t size)
 {
 	ARG_UNUSED(dummy);
 
+#if defined(CONFIG_CONSOLE_RX_INTERRUPT_DRIVEN)
```

Comment:
I wonder if this could use if() instead of #ifdef?

Reply:
That is not possible because `console_serial` accessed below is conditionally compiled in

Reply:
So if you use if () and the condition evals to 0, doesn't the compiler remove console_serial automatically? If so, then you don't need the #ifdef there either.

It is possible that the compiler doesn't handle this, but in my experience it normally works. I'm assuming we are using -fdata-sections etc.

Reply:
I tried. It does not compile